### DISCUSSION TITLE
Setting up a WiFi Breath Controller (ESP32)

### DISCUSSION THREAD

**Discussion by KungFuHaiku** [tttapa/Control-Surface#758] [Q&A] [answered]
Hi Folks,

I was wondering if anyone knew how I might go about setting up a MIDI breath controller using this library.  I've already managed to set up a simple WiFI transport controller using its AppleMIDI component (start, stop, play, record), but I'm unsure of which function I need to use to send breath control messages.

I'm using the HX710B pressure sensor, and have found an Arduino library that claims to be able to plot its output.

HX710B Sensor:
https://sharvielectronics.com/product/hx710b-air-pressure-0-40kpa-sensor-module/

HX710B Library:
https://github.com/kurimawxx00/hx710B_pressure_sensor

Many thanks.

**Comment by tttapa** [comment, marked as answer]
Depends on the kind of MIDI message you want to send, but this one comes to mind:
```cpp
// 0 ≤ value < 128
Control_Surface.sendControlChange(MIDI_CC::Breath_Controller, value);
```
See also:
- https://tttapa.github.io/Control-Surface-doc/Doxygen/d3/df7/midi-tutorial.html#midi_md-sending
- https://tttapa.github.io/Control-Surface-doc/Doxygen/d6/d72/classMIDI__Sender.html

**Comment by KungFuHaiku** [comment]
That's great, thank you.  I was going to try and avoid adding anything within the loop, but that should work fine.  Out of interest, is there a MIDI Output Element for breath control (like CCPotentiometer, that can be instantiated outside the loop)?

Thanks again.

**Reply by tttapa** [reply]
You could write a custom MIDI Output Element for it: https://tttapa.github.io/Control-Surface-doc/Doxygen/d4/d3b/Custom-MIDI-Output-Element_8ino-example.html

```cpp
#include <Control_Surface.h>
#include <AH/STL/algorithm>
#include <AH/STL/limits>

uint16_t breathcontrol_read(pin_t p) {
  return analogRead(p);  // simple example
}

class CCBreathController : public MIDIOutputElement {
  public:
    CCBreathController(pin_t pin, MIDIChannelCable ch)
      : pin(pin), ch(ch) {}

    using ValueType = uint_fast16_t; // type of the measured values
    using FilterType = uint_fast16_t; // type for the filter state
    constexpr static uint8_t filter_shift_bits = 3; // amount of filtering
    constexpr static uint8_t sensor_bits = ADC_BITS; // precision of the sensor
    constexpr static uint8_t final_bits = 7; // desired precision of the output
    // How many bits do we have left for additional precision in intermediate results?
    constexpr static uint8_t inc_bits =
      std::min(std::numeric_limits<FilterType>::digits - sensor_bits - filter_shift_bits,
               std::numeric_limits<ValueType>::digits - sensor_bits);

    // This method is called once by `Control_Surface.begin()`.
    void begin() final override;
    // This method is called continuously by `Control_Surface.loop()`.
    void update() final override;

    ValueType getRawValue() const;
    constexpr static ValueType getMinRawValue() {
      return ValueType {0};
    }
    constexpr static ValueType getMaxRawValue() {
      static_assert(std::is_unsigned<ValueType>::value, "Fix this for signed types");
      return (ValueType {1} << (sensor_bits + inc_bits)) - ValueType {1};
    }

  private:
    pin_t pin;
    MIDIChannelCable ch;
    using EMA_t = EMA<filter_shift_bits, ValueType, FilterType>;

    EMA_t filter;
    Hysteresis<sensor_bits + inc_bits - final_bits, ValueType, ValueType> hysteresis;
};

void CCBreathController::begin() {
  static_assert(EMA_t::supports_range(getMinRawValue(), getMaxRawValue()),
                "Error: EMA filter type doesn't support full ADC range");
  auto widevalue = getRawValue();
  filter.reset(widevalue);
  hysteresis.setValue(widevalue);
}

void CCBreathController::update() {
  ValueType input = getRawValue();  // read the raw sensor value
  input = filter.filter(input);     // apply a low-pass EMA filter
  input = -input;                   // optionally transform the measurement
  if (hysteresis.update(input))  // if the value changed since last time
    Control_Surface.sendControlChange({MIDI_CC::Breath_Controller, ch}, 
                                      hysteresis.getValue());
}

CCBreathController::ValueType CCBreathController::getRawValue() const {
  ValueType value = breathcontrol_read(pin);
  return increaseBitDepth<sensor_bits + inc_bits, sensor_bits, ValueType>(value);
}

// :::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::: //

USBDebugMIDI_Interface midi;

CCBreathController bc { A0, CHANNEL_1 };

void setup() {
  Control_Surface.begin(); // Initialize Control Surface (calls button.begin())
}

void loop() {
  Control_Surface.loop(); // Update the Control Surface (calls button.update())
}
```

**Reply by KungFuHaiku** [reply]
I've managed to use the Q2HX711 library to retrieve the value of the pressure sensor, along with converting it to a range that's between 0 and 127.  However, I can't work out how to adapt your code to use this value.  I don't need to pass any pins into the function, and I suspect I don't need any of the hysteresis, filtering or limiting functionality.  I hope it's not too cheeky, but could I perhaps ask for some pointers to simplify this, so it can just accept a 0-127 value (whilst keeping everything outside of the loop)?

Thanks again!

**Comment by KungFuHaiku** [comment]
That's fantastic - thanks very much for taking the time to provide some code.  I'm just waiting on a couple of hardware components, but I'll be sure to post an update on my progress.

Many thanks.
